gears/src/crypto/public.rs

```diff
@@ -22,12 +30,77 @@ impl PublicKey {
     ) -> Result<(), SigningError> {
         match self {
             PublicKey::Secp256k1(key) => key.verify_signature(message, signature),
+            PublicKey::Ed25519(key) => key.verify_signature(message, signature),
+        }
+    }
+
+    pub fn get_address(&self) -> AccAddress {
+        match self {
+            PublicKey::Secp256k1(key) => key.get_address(),
+            PublicKey::Ed25519(key) => key.get_address(),
+        }
+    }
+}
+
+impl TryFrom<Any> for PublicKey {
+    type Error = DecodeError;
+
+    fn try_from(any: Any) -> Result<Self, Self::Error> {
+        match any.type_url.as_str() {
+            "/cosmos.crypto.secp256k1.PubKey" => {
+                let key = Secp256k1PubKey::decode::<Bytes>(any.value.into())
+                    .map_err(|e| DecodeError(e.to_string()))?;
+                Ok(Self::Secp256k1(key))
+            }
+            "/cosmos.crypto.ed25519.PubKey" => {
+                let key = Ed25519PubKey::decode::<Bytes>(any.value.into())
+                    .map_err(|e| DecodeError(e.to_string()))?;
+                Ok(Self::Ed25519(key))
+            }
+
+            _ => Err(DecodeError(format!(
+                "Key type not recognized: {}",
+                any.type_url
+            ))),
         }
     }
 }
 
-impl From<Secp256k1PubKey> for PublicKey {
-    fn from(value: Secp256k1PubKey) -> Self {
-        Self::Secp256k1(value)
+impl From<PublicKey> for Any {
+    fn from(key: PublicKey) -> Self {
+        match key {
+            PublicKey::Secp256k1(key) => Any {
+                type_url: "/cosmos.crypto.secp256k1.PubKey".to_string(),
+                value: key.encode_vec(),
+            },
+            PublicKey::Ed25519(key) => Any {
+                type_url: "/cosmos.crypto.ed25519.PubKey".to_string(),
+                value: key.encode_vec(),
+            },
+        }
+    }
+}
+
+use tendermint::types::proto::crypto::PublicKey as TendermintPublicKey;
```

Comment:
Interesting that for this you didn't follow my way with `inner` mod.

Reply:
I think the use case is a bit different. We usually use inner because the user doesn't need to know about that type. In this case users are aware of the type - we're giving them From and Into methods for the type.

Reply:
Not sure that this is 100% correct, 'cause you don't re-export this type. And `inner` for imports is not requirements. Just a way which easiest for me and I noted that you and @dmvict  follow it.